UI Next workflow graph edit/results

mabashian opened the pull request:
##### SUMMARY
With this PR, a user can now edit workflow graphs in UI Next.  There are a couple of things missing from this implementation which I will outline:

1) Prompting for JT nodes that allow it - this will come later
2) "View" mode on the nodes - work on this feature is already in progress but I decided to peel it out into a separate follow-on PR

This PR also builds on https://github.com/ansible/awx/pull/5515 (which I will close since it never really gained traction).  Users can view the output graph for a workflow job with this pr (however, no websocket/live update support at this time).

Known bugs/todos:

1) The start node is not dynamically generated so it will not scale with translated strings.  I plan on creating a followup issue for this.
2) Tests - will be added to this PR as I have time to work on them.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

mabashian committed:
Render workflow results.  Extensive refactors of workflow components in general.

mabashian committed:
Fully functioning workflow editor without read-only view modal and without prompting.

mabashian committed:
Halfway implemented node details.  Still need to handle cases where the user has edited the node and cases where the node is brand new.

mabashian committed:
Stabilized workflow visualizer and output point.  Workflow jobs can be viewed and workflows can be built (without jt prompting).

mabashian committed:
Prettier

AlexSCorey commented on awx/ui_next/src/components/CodeMirrorInput/VariablesDetail.jsx:
```diff
@@ -9,7 +9,7 @@ import { yamlToJson, jsonToYaml, isJson } from '../../util/yaml';
 const JSON_MODE = 'javascript';
 
-function VariablesDetail({ value, label, rows }) {
+function VariablesDetail({ value = '---', label, rows }) {
```
You will probably need to rebase this commit to get my pr from yesterday as I introduced code that does the same thing your value does above. I just did it here https://github.com/ansible/awx/blob/devel/awx/ui_next/src/components/CodeMirrorInput/VariablesDetail.jsx#L61

AlexSCorey commented:
![Workflow Bug](https://user-images.githubusercontent.com/39280967/72822789-af02ef00-3c40-11ea-8c60-adeb42f0f14e.gif)
When I remove a link, regardless of how many levels down it is that Job Template is added to the first level.  Deleting the link does not seem to be a feature in the old ui.

AlexSCorey commented:
<img width="642" alt="Screen Shot 2020-01-21 at 11 37 08 AM" src="https://user-images.githubusercontent.com/39280967/72823920-6ea47080-3c42-11ea-98fd-66bd93bc6ea9.png">
It seems like these should close when you click outside of them on the screen, but they don't.  You have to click on the icon that opened them to get them to close.

EDIT:  I do see the value in the Key "legend" remaining open even if the user clicks outside of it.  But the Toggle Tools should still close when you click outside of it I believe.

AlexSCorey commented:
@mabashian @trahman73 What do you think about allowing the user to turn off the tooltip that appears in the upper left hand corner.  I found myself getting annoyed with it constantly popping up as my mouse moved across the screen, however as a user I wouldn't want it to be remove all together.

trahman73 commented:
@AlexSCorey Yup, that's the longer term plan for that area. We would have it work like a toggle similar to Key. Also, we do want to add a close icon to the Key and Tools panels.

AlexSCorey commented on awx/ui_next/src/components/Workflow/WorkflowActionTooltip.jsx:
```diff
@@ -0,0 +1,69 @@
+  display: flex;
+`;
+
+const TooltipArrows = styled.div`
```
Is it possible for a Tooltip to have more than 1 arrow?  If not, should this be singular?

AlexSCorey commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModal/NodeTypeStep/JobTemplatesList.jsx:
```diff
@@ -0,0 +1,87 @@
+  order_by: 'name',
+});
+
+function JobTemplatesList({ i18n, history, nodeResource, updateNodeResource }) {
```
Should `updateNodeResource` be `onUpdateNodeResource` since it's a function being passed in as a prop?

AlexSCorey commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModal/NodeTypeStep/NodeTypeStep.jsx:
```diff
@@ -0,0 +1,271 @@
+  timeout,
+  updateDescription,
+  updateName,
+  updateNodeResource,
```
Should `updateNodeResource` be `onUpdateNodeResource` since it's a function being passed in as a prop? As well as the same pattern for `updateDescription` being `onUpdateDescription`?

AlexSCorey commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModal/NodeTypeStep/InventorySourcesList.jsx:
```diff
@@ -0,0 +1,90 @@
+          key={item.id}
+          name={item.name}
+          label={item.name}
+          onSelect={() => updateNodeResource(item)}
```
Should updateNodeResource be onUpdateNodeResource since it's a function being passed in as a prop?

AlexSCorey commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/VisualizerGraph.jsx:
```diff
@@ -0,0 +1,369 @@
+                    onLinkEditClick={onLinkEditClick}
+                    readOnly={readOnly}
+                    updateHelpText={setHelpText}
+                    updateLinkHelp={setLinkHelp}
```
These props should be `onUpdate.....`

AlexSCorey commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/VisualizerGraph.jsx:
```diff
@@ -0,0 +1,369 @@
+                    onViewNodeClick={onViewNodeClick}
+                    readOnly={readOnly}
+                    updateHelpText={setHelpText}
+                    updateNodeHelp={setNodeHelp}
```
Same thing here

mabashian committed:
Adds unit test coverage for shared workflow components

mabashian committed:
Update function/callback props passed into workflow components to start with "on"

mabashian committed:
Default '---' not needed due to changes to how the prop is passed to CodeMirrorInput

mabashian committed:
Add close button to workflow tools/key

jlmitch5 commented on awx/ui_next/src/screens/Job/WorkflowOutput/WorkflowOutput.jsx:
```diff
@@ -0,0 +1,229 @@
+  }, [job.id, job.unified_job_template, i18n]);
+
+  // Update positions of nodes/links
+  useEffect(() => {
```
this is a nice separation of concerns between data (above) and view (below).  I like the way you did this a lot, it's very easy to follow.

jlmitch5 commented on awx/ui_next/src/screens/Job/WorkflowOutput/WorkflowOutputStartNode.jsx:
```diff
@@ -0,0 +1,28 @@
+import { shape } from 'prop-types';
+import { constants as wfConstants } from '@util/workflow';
+
+function WorkflowOutputStartNode({ nodePositions }) {
```
It seems a bit strange this utilizes the vanilla svg elements, but the WorkflowOutputNode uses the component wrappers.

jlmitch5 commented on awx/ui_next/src/screens/Job/WorkflowOutput/WorkflowOutputToolbar.jsx:
```diff
@@ -0,0 +1,90 @@
+  const totalNodes = nodes.reduce((n, node) => n + !node.isDeleted, 0) - 1;
+
+  return (
+    <div css="border-bottom: 1px solid grey; height: 56px; display: flex; alignItems: center">
```
I would move all these elements that use the css= attr to using styled components

jlmitch5 commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModal/NodeTypeStep/InventorySourcesList.jsx:
```diff
@@ -0,0 +1,90 @@
+      items={inventorySources}
+      onRowClick={row => onUpdateNodeResource(row)}
+      qsConfig={QS_CONFIG}
+      showPageSizeOptions={false}
```
do we want the inventory search and sort columns here?

mabashian replied:
👍 yea this list is for inventory sources so will differ a bit from the inventory list.  I don't think we've built the inventory sources list quite yet so I'll just add some generic search/sort columns until we nail down the fields we want for that list.

jlmitch5 commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModal/NodeTypeStep/JobTemplatesList.jsx:
```diff
@@ -0,0 +1,92 @@
+      itemCount={count}
+      items={jobTemplates}
+      onRowClick={row => onUpdateNodeResource(row)}
+      qsConfig={QS_CONFIG}
```
Same here about the search and sort columns

jlmitch5 commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModal/NodeTypeStep/ProjectsList.jsx:
```diff
@@ -0,0 +1,85 @@
+      itemCount={count}
+      items={projects}
+      onRowClick={row => onUpdateNodeResource(row)}
+      qsConfig={QS_CONFIG}
```
search/sortColumns?

jlmitch5 commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModal/NodeTypeStep/WorkflowJobTemplatesList.jsx:
```diff
@@ -0,0 +1,92 @@
+      itemCount={count}
+      items={workflowJobTemplates}
+      onRowClick={row => onUpdateNodeResource(row)}
+      qsConfig={QS_CONFIG}
```
search/sortColumns?

jlmitch5 commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Visualizer.jsx:
```diff
@@ -187,19 +797,125 @@ function Visualizer({ template, i18n }) {
+  template: shape().isRequired,
+};
+
+export default withI18n()(withRouter(Visualizer));
```
There seems to be a lot of duplication between WorkflowOutput/WorkflowOutputGraph and Visualizer/VisualizerGraph, as well as the link/node/start node/start screen, toolbar and help  In an ideal world, it would be nice if you could use the same backbone and put the "form-on-click" kind of stuff in here.  Not something I think we need to change for this PR, but the duplication does seem like it may cause change drift between view and edit over time.

jlmitch5 commented on awx/ui_next/src/util/workflow.jsx:
```diff
@@ -0,0 +1,204 @@
+    (newScale * svgBoundingClientRect.height - svgBoundingClientRect.height) /
+      2;
+  return [translateX, translateY];
+}
```
I get why we are doing this (share between the output and edit graphs), but it feels a bit weird in utils.

For the future it would be nice if all the graph-related stuff lived in components/workflows and that's the bulk of all the code and then the stuff in screens between Visualizer and Output were ver short files that maybe instatiate that combine the api endpoints, call the workflow graph components and maybe instatiate a prop that makes the graph act as edit/view.

mabashian replied:
I moved this into components/Workflow

jlmitch5 reviewed (changes requested):
looking great!

mabashian committed:
Adds unit test coverage to some of the workflow output components.  Also adds support for hovering on workflow results links to see the edge type (success/fail/always).

dsesami commented:
Using this comment as a running list of suggestions and issues:

- Is it possible to make it so that the "active field" for keeping the submenu open a bit larger? For example, if I follow this path with my mouse to try to click the +, the submenu vanishes before I can get there:
![Screenshot from 2020-01-23 14-55-01](https://user-images.githubusercontent.com/19942822/73018980-ab15cf00-3df0-11ea-9972-a09bc27ac784.png)

- Also, centering on the node type is a bit off:
![Screenshot from 2020-01-23 15-01-07](https://user-images.githubusercontent.com/19942822/73019296-4e66e400-3df1-11ea-8c89-32aa08a67dee.png)

- Personally, I don't think it's really obvious that the buttons in this menu are buttons (the fit to screen button, the + and - signs specifically). They could use some light borders or shading or something.
![Screenshot from 2020-01-23 15-12-07](https://user-images.githubusercontent.com/19942822/73020058-d994a980-3df2-11ea-8944-c18c658a9879.png)

- Can we call the Key a Legend instead? it works a little better since the symbol is a compass.

- When hovering over greyed-out items in the top bar, there is no tooltip to indicate what these symbols mean. The tooltip should still work to inform the user.
![Screenshot from 2020-01-23 15-18-30](https://user-images.githubusercontent.com/19942822/73020493-b9191f00-3df3-11ea-8413-26070a41689a.png)

- I feel like this could be a bit more explicit. "Exit Without Saving" or something like that.
![Screenshot from 2020-01-23 15-20-58](https://user-images.githubusercontent.com/19942822/73020636-0ac1a980-3df4-11ea-9ba9-4dc57449f4d6.png)

- Attempting to save the workflow resulted in this message. Is this known?
```Uncaught (in promise) TypeError: Cannot read property 'summary_fields' of undefined
    at eval (Visualizer.jsx:473)
    at Array.forEach (<anonymous>)
    at handleVisualizerSave (Visualizer.jsx:417)
    at HTMLUnknownElement.callCallback (react-dom.development.js:362)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:411)
    at invokeGuardedCallback (react-dom.development.js:466)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:480)
    at executeDispatch (react-dom.development.js:612)
    at executeDispatchesInOrder (react-dom.development.js:637)
    at executeDispatchesAndRelease (react-dom.development.js:743)```

jakemcdermott commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModal/NodeNextButton.jsx:
```diff
@@ -0,0 +1,39 @@
+  triggerNext,
+}) {
+  useEffect(() => {
+    if (!triggerNext) {
```
I traced how click events are handled between this component and its parent, `NodeModal` and as far as I can tell it works like this:

1. `NodeNexButton` is clicked, which calls `onClick(activeStep)`
2. The implementation of `onClick` lives in the parent component, `NodeModal`. It doesn't  accept any arguments though, so `activeStep` is ignored(?)
3. `onClick` updates a state variable, `triggerNext`
4. `triggerNext` is passed _back_ down into `NodeNextButton` as a prop
5. The `useEffect` hook inside of `NodeNextButton` calls `onNext` as a side-effect of `triggerNext` changing, so long as `triggerNext` doesn't equal 0. 
6. `onNext`  then calls back up to... the parent, `NodeModal`. This function call is what actually transitions the component to the next page.

This seems like a _lot_ of steps and events propagating back and forth between the components just to handle the `next` button click. I'm having trouble seeing why the `triggerNext` variable is needed, and I'm wondering if I'm just not understanding the problem it is meant to solve.

Just to see what would happen, I bypassed steps 2,3,4, and 5  by removing the `useEffect` hook entirely, deleting the `triggerNext` variable from the `NodeButton` component,  and just passing `onNext` directly to the button as its click handler. As far as I can tell everything still works? :confused:

mabashian replied:
@jakemcdermott an astute observation.  I worked on this component with @keithjgrant before the break I believe.  This component is in place to solve a single problem: Adding steps to the workflow wizard dynamically.  In a previous iteration of this PR, the wizard steps were a bit different:

Step 1: Select the edge type
Step 2: Select the node type
Step 3: (Depended on what you selected from step 3 and is conditionally included in the steps array based on a state variable.)

The problem that we were running in to was that when the user clicked next, the wizard's `onNext` function was getting called _before_ the state variable was through updating (and steps array subsequently updated).  So the user would not get navigated to the dynamically added 3rd step but rather, the wizard would close.

This isn't immediately apparent the way the PR currently sits because the UX has changed slightly and we no longer dynamically add that 3rd step but it has been combined with step 2.  I didn't remove this component however because I think we will need this behavior as soon as we add prompting in.  The prompting steps will need to be dynamically added and I think we'll run into the same problem that this component solves.

keithjgrant commented on awx/ui_next/jest.config.js:
```diff
@@ -11,6 +11,7 @@ module.exports = {
     '\\.(css|scss|less)$': '<rootDir>/__mocks__/styleMock.js',
     '^@api(.*)$': '<rootDir>/src/api$1',
     '^@components(.*)$': '<rootDir>/src/components$1',
+    '@constants$': '<rootDir>/src/constants.js',
```
Should this start with `^` like the other path shortcuts?

mabashian replied:
Yea - it's safer that way

mabashian replied:
Should now be fixed

mabashian committed:
Moves inline css to styled component in WorkflowOutputToolbar

mabashian committed:
Adds default search and sort columns to the four lists in the workflow node wizard

mabashian committed:
Combines the two start node components into one.  Removes use of document.getElementById in workflow components in favor of refs.

mabashian committed:
Moves util/workflow.jsx to components/Workflow/WorkflowUtils.jsx and updates imports

keithjgrant commented:
Opening the "i" info modal on a deleted resource node crashes with an uncaught error

![wf-visualizer-bug](https://user-images.githubusercontent.com/410794/73093293-0a2c2000-3e93-11ea-8773-e312b2e1b8d3.gif)

keithjgrant commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Visualizer.jsx:
```diff
@@ -1,94 +1,688 @@
 };
 
-function Visualizer({ template, i18n }) {
+function Visualizer({ history, template, i18n }) {
```
This component is huge. I would look for ways to break it up. Your best bet might be to put a lot of this state into a reducer and add a `useReducer` to manage changes and extract a lot of this state logic into smaller functions.

keithjgrant commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Visualizer.jsx:
```diff
@@ -1,94 +1,688 @@
+    setNodeToEdit(null);
+  };
+
+  const deleteNode = () => {
```
This function is pretty long. I'd consider moving chunks of it into helper functions to it's easier to mentally track everything that's going on... probably lines 165–183 into one function ("Remove any links that reference this node") and 185–208 ("Add the new links")

keithjgrant commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Visualizer.jsx:
```diff
@@ -1,94 +1,688 @@
+    // Find and mark any ancestors as disabled to prevent cycles
+    links.forEach(link => {
+      // id=1 is our artificial root node so we don't care about that
+      if (link.source.id !== 1) {
```
Consider checking for the opposite and returning early:

```
if (link.source.id === 1) {
  return;
}
```

This prevents nested `if`s and simplifies the logic a little

mabashian replied:
👍 on it

mabashian commented:
@dsesami - see replies below

> * Is it possible to make it so that the "active field" for keeping the submenu open a bit larger? For example, if I follow this path with my mouse to try to click the +, the submenu vanishes before I can get there:

Yea, I agree.  I've got a few ideas for how to solve this but each with their own drawbacks.  If it's alright with you I'd like peel this off into a follow-on issue.

> * Also, centering on the node type is a bit off:

Yea, the default vertical align on the icon components is throwing that one off.  I'll see what I can do.

> * Personally, I don't think it's really obvious that the buttons in this menu are buttons (the fit to screen button, the + and - signs specifically). They could use some light borders or shading or something.

Talked with Taufique about this one - I think we're going to change all of those buttons to pf buttons with tooltips.  This'll make the tools a bit larger but hopefully make it more apparent that they are buttons to be clicked.

> * Can we call the Key a Legend instead? it works a little better since the symbol is a compass.

Sure

> * When hovering over greyed-out items in the top bar, there is no tooltip to indicate what these symbols mean. The tooltip should still work to inform the user.

Yea this is a PF thing - see: https://github.com/patternfly/patternfly-react/issues/1894

> * I feel like this could be a bit more explicit. "Exit Without Saving" or something like that.

Sure

> * Attempting to save the workflow resulted in this message. Is this known?

This is not known to me.  I may need you to help me with some reproducer steps.

keithjgrant commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Visualizer.jsx:
```diff
@@ -1,94 +1,688 @@
+    }
+  };
+
+  const handleVisualizerSave = async () => {
```
450 line function. This would be much easier to follow once it's broken up into smaller pieces. Breaking it up should also help simplify a lot of the complex/nested if-else logic in the first section.

For example, down near the end, lines 518–650 could be simplified to this, with the looping & bare logic relegated to smaller functions:

```
const {linkMap, newLinks} = buildNewLinkMap(links, originalLinkMap);
const disassociateRequests = disassociateDeletedLinks(linkMap, originalLinkMap, deletedNodeIds);
await Promise.all(disassociateRequests);

const associateRequests = associateNewLinks(newLinks);
await Promise.all(associateRequests);
history.push(`/templates/workflow_job_template/${template.id}/details`);
```

If you apply this sort of treatment to the whole function, I think you could probably get it down to maybe 50 lines or so. Then it's much easier to understand at a glance what's going on in the saving process.

mabashian committed:
Change Exit button text to Exit Without Saving in the unsaved changes modal.

mabashian committed:
Vertically center the pause icon in the node type letter component

mabashian committed:
Turns all the workflow tool buttons in to pf button components.

mabashian committed:
Changes "Key" references to "Legend"

mabashian committed:
Run prettier

keithjgrant commented on awx/ui_next/src/screens/Job/WorkflowOutput/WorkflowOutput.jsx:
```diff
@@ -0,0 +1,229 @@
+  const [showLegend, setShowLegend] = useState(false);
+  const [showTools, setShowTools] = useState(false);
+
+  useEffect(() => {
```
another very long function

mabashian committed:
Properly bookend @constants alias so that it doesn't inadvertently match something we don't want it to match

mabashian committed:
Removes reference to the node unified job template name in the view modal for now.  This component is really just a placeholder and this change fixes an error that is thrown when the node's unified job template is deleted.

mabashian commented:
re: https://github.com/ansible/awx/pull/5712#issuecomment-578242915 - that component is really just a placeholder until I can finish the real modal (after this PR).  I went ahead and just removed the ujt name from the title of the modal.  That should take care of the error for now.

mabashian committed:
Move visualizer/workflow output state logic out to reducer and refactor some of the larger functions.  Introduces contexts for state/dispatch that can be used by descendent components of both the visualizer and the workflow output components.

mabashian committed:
Fix some styling discrepancies between Chrome and Firefox in the workflow viz/output graphs.

Cleans up deleted job/job template use cases.  Show message indicating that the ujt associated with a node has been deleted.

marshmalien commented on awx/ui_next/src/components/HorizontalSeparator/HorizontalSeparator.jsx:
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
```
PatternFly recently promoted the [Divider component](https://github.com/patternfly/patternfly-react/commits/master/packages/patternfly-4/react-core/src/components/Divider/Divider.tsx) from experimental. Just wanted to put this on your radar.

marshmalien commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModals/NodeModal.jsx:
```diff
@@ -0,0 +1,213 @@
+  title: node.isRequired,
+};
+
+export default withI18n()(withRouter(NodeModal));
```
Instead of using `withRouter`, you could pull in react-router hooks `let history = useHistory();`.

mabashian replied:
👍 there were a few other places where I was using withRouter.  I'll replace those as well.

marshmalien commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModals/NodeModal.jsx:
```diff
@@ -0,0 +1,213 @@
+        param
+      )
+    );
+    history.push(`${history.location.pathname}?${otherParts.join('&')}`);
```
Whenever I add a new node, this error shows up in the console: 
<img width="676" alt="Screen Shot 2020-01-28 at 5 05 12 PM" src="https://user-images.githubusercontent.com/15881645/73311094-afc4f380-41f3-11ea-9dee-1ca649a71757.png">

You could swap history.push for history.replace to get around this - it replaces the current entry on the history stack.

marshmalien commented on awx/ui_next/src/screens/Job/Job.jsx:
```diff
@@ -124,12 +126,24 @@ class Job extends Component {
-                render={() => <JobOutput type={match.params.type} job={job} />}
+                render={() =>
+                  match.params.type === 'workflow' ? (
+                    <WorkflowOutput job={job} />
```
The job getting passed down to WorkflowOutput is stale bc we are trying fetch a workflow with the previous job obj. 
![wf](https://user-images.githubusercontent.com/15881645/73373105-388c6f80-4286-11ea-98b9-58eaff6f51d4.gif)

marshmalien commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModals/NodeTypeStep/NodeTypeStep.jsx:
```diff
@@ -0,0 +1,271 @@
+                              field.onChange(evt);
+                            }}
+                          />
+                          <TimeoutLabel>min</TimeoutLabel>
```
Do we need to mark the timeout labels for internationalization?

keithjgrant commented on awx/ui_next/src/components/Workflow/workflowReducer.js:
```diff
@@ -0,0 +1,632 @@
+
+  // Filter out the duplicates
+  invalidLinkTargetIds
+    .filter((element, index, array) => index === array.indexOf(element))
```
I think this check will always return true, so this `filter` has no effect

mabashian replied:
This filters the duplicate node id's out of the initial array.  

<img width="593" alt="Screen Shot 2020-01-30 at 1 23 25 PM" src="https://user-images.githubusercontent.com/9889020/73478234-ba9e9600-4363-11ea-9a3e-0be83018a4aa.png">

While not 100% necessary this potentially reduces the amount of looping we have to do in the next line.

mabashian committed:
Adds test coverage to the workflow output and workflow output graph components

mabashian committed:
Replace withRouter with useHistory hook

mabashian committed:
Swap history.push for history.replace to alleviate console warning dealing with pushing to the same path.

mabashian committed:
Mark min/sec labels for translation

mabashian committed:
Fixes bug where viewing the workflow output, clicking a node to view the details of that job and then hitting the back button would result in an error and no output graph.

mabashian committed:
Fixes bug where navigating from one output route to another was breaking

mabashian committed:
Removes unnecessary 'type' field on nodes and links

mabashian committed:
Adds unit test coverage for workflow reducer

mabashian committed:
Adds basic unit test coverage to visualizer components (not including modals).

mabashian committed:
Remove HorizontalSeparator component.  Patternfly now has a Divider component that we can use if we need something like this.

mabashian committed:
Adds tests for delete all nodes and unsaved changes modals

mabashian committed:
Adds test for link add/edit/delete modals

unlikelyzero commented:
@mabashian some feedback:

- I am unable to save a WFJT after adding a JT which is missing an inventory. I start with an empty WFJT and add a single JT. I get the following console error
`createError.js:16 Uncaught (in promise) Error: Request failed with status code 400
    at createError (createError.js:16)
    at settle (settle.js:18)
    at XMLHttpRequest.handleLoad (xhr.js:59)`
with a HTTP 400 API Response `{“resources_needed_to_start”:[“Job Template project is missing or undefined.“,”Job Template inventory is missing or undefined.“]}`

- I will need to work out some more IDs and testcase stubs
--- Toolbar
--- Launch Button
--- Delete Button
--- Save Button
--- Start Button

@trahman73 , the tab pop-out to fullscreen workflow visualizer is jaring. Is there a UX pattern to indicate that the tab click will lead to a fullscreen editor? Something like 
<img width="41" alt="Screen Shot 2020-02-05 at 12 54 13 PM" src="https://user-images.githubusercontent.com/12446869/73868544-a7ce0a80-4816-11ea-8606-dc3fc234d0d7.png">

trahman73 commented:
@unlikelyzero We can look into adding a similar PF icon that's passed into the PF tab component. I see they have an example of it here: https://www.patternfly.org/v4/documentation/react/components/tabs

mabashian committed:
Adds test coverage for node modal components

mabashian committed:
Adds ID's to toolbar(s) and action buttons within visualizer/output toolbars.

mabashian committed:
Adds id to workflow visualizer start screen button

mabashian committed:
Run prettier

mabashian committed:
Adds tests for the node type step components in the node modal

mabashian committed:
Hide message stating that the resource has been deleted if the output node is an approval node

mabashian committed:
Disable tools/legend buttons when the workflow has no nodes and is displaying the start screen.

Commit:
Add IDs for nodes badge and template name

mabashian committed:
Run prettier to fix failures

mabashian committed:
Remove Workflow Visualizer text from visualizer toolbar

mabashian committed:
Adds id's to the node name and the node type letter

unlikelyzero reviewed (approved):
Tested and have a PR for the tests here:
https://github.com/ansible/tower-qa/pull/4668

Low Priority Issues filed ( if not fixed):
https://github.com/ansible/awx/issues/5951
https://github.com/ansible/awx/issues/5952
https://github.com/ansible/awx/issues/5953
